lib/galaxy/managers/cloud.py

```diff
@@ -0,0 +1,148 @@
+
+            config = {'aws_access_key': access,
+                      'aws_secret_key': secret}
+            connection = CloudProviderFactory().create_provider(ProviderList.AWS, config)
+        elif provider == "azure":
+            subscription = credentials.get('subscription_id', None)
+            if subscription is None:
+                missing_credentials.append('subscription_id')
+            client = credentials.get('client_id', None)
+            if client is None:
+                missing_credentials.append('client_id')
+            secret = credentials.get('secret', None)
+            if secret is None:
+                missing_credentials.append('secret')
+            tenant = credentials.get('tenant', None)
+            if tenant is None:
+                missing_credentials.append('tenant')
+            if len(missing_credentials) > 0:
+                raise RequestParameterMissingException("The following required key(s) are missing from the provided "
+                                                       "credentials object: {}".format(missing_credentials))
+
+            config = {'azure_subscription_id': subscription,
+                      'azure_client_id': client,
+                      'azure_secret': secret,
+                      'azure_tenant': tenant}
+            connection = CloudProviderFactory().create_provider(ProviderList.AZURE, config)
+        else:
+            raise RequestParameterInvalidException("Unrecognized provider '{}'; the following are the supported "
+                                                   "providers: {}.".format(provider, SUPPORTED_PROVIDERS))
+
+        try:
+            if connection.authenticate():
+                return connection
+        except ProviderConnectionException as e:
+            raise AuthenticationFailed("Could not authenticate to the '{}' provider. {}".format(provider, e))
+
+    def download(self, trans, history_id, provider, container, obj, credentials):
+        if CloudProviderFactory is None:
+            raise Exception(NO_CLOUDBRIDGE_ERROR_MESSAGE)
+
+        connection = self._configure_provider(provider, credentials)
+        try:
+            container_obj = connection.object_store.get(container)
+            if container_obj is None:
+                raise RequestParameterInvalidException("The container `{}` not found.".format(container))
+        except Exception as e:
+            raise ItemAccessibilityException("Could not get the container `{}`: {}".format(container, str(e)))
+
+        key = container_obj.get(obj)
+        if key is None:
+            raise ObjectNotFound("Could not get the object `{}`.".format(obj))
+        staging_file_name = os.path.abspath(os.path.join(
+            trans.app.config.new_file_path,
+            "cd_" + ''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(11))))
+        staging_file = open(staging_file_name, "w+")
```

Comment:
Looks to me like the `close` at some point is missing.

Reply:
I was mainly relying on an auto-close upon leaving the scope of 'open'. However, to make sure the handle is closed, now I call `close` at the end of the function.

Reply:
The best way to generally do this is the python `with` statement.  See http://www.pythonforbeginners.com/files/with-statement-in-python for more info.

Reply:
That is correct. However, since there is another `with` with a long body, I decided to use `close` to avoid nested withs.

Reply:
I don't follow the concern, `with`s can be nested and it's the safest way to ensure closure, but as long as you're correctly closing it doesn't matter too much.